### ISSUE TITLE
evdev clients failing to attach

### ISSUE THREAD

**Issue by djvs** [kaiwang27/uhidd#10]
Tl:dr: uhidd will detach the keyboard when -u is specified, will log the keypresses if -vv if specified, and will attempt to dispatch them to evdev and vkbd, but only vkbd works, apparently because there no evdev clients that get registered.

------------

Not really sure how the internals of this library work.  Been dumping printf statements out and such trying to debug why ` ./uhidd -ukosvv /dev/ugen0.2 /usr/local/etc/uhidd.conf` is dumping keycodes to TTY but not to X/evdev.  This slightly modified code:

``` 355 static int
    356 evdev_alloc_cuse_dev(struct evdev_dev *ed)
-   357 {
|   358     int classid, devid, i;
|   359
|   360     if (ucuse_init() < 0)
|   361         return (-1);
|   362
|   363     classid = EVDEV_CUSE_INDEX - 'A';
|   364     if (cuse_alloc_unit_number_by_id(&devid, CUSE_ID_UHIDD(classid)) !=
-   365         0) {
2   366         syslog(LOG_ERR, "cuse_alloc_unit_number_by_id failed");
2   367         return (-1);
2   368     }
|   369
|   370     printf("cuse_dev_create here %u %s\n", devid, EVDEV_CUSE_DEFAULT_DEVNAME);
|   371     printf("cuse_dev result %p \n", cuse_dev_create(&evdev_cuse_methods, ed, NULL, 0, 0, 0660, "%s%d", EVDEV_CUSE_DEFAULT_D|       EVNAME, devid));
|   372
|   373
|   374     snprintf(ed->devname, sizeof(ed->devname), "%s%d", EVDEV_CUSE_DEFAULT_DEVNAME, devid);
|   375
|   376     /* Create worker threads. */
|   377     for (i = 0; i < 2; i++)
|   378         ucuse_create_worker();
|   379
|   380     return (0);
|   381 }
```

Dumps a 0x0 pointer for the return value of cuse_dev_create.  No idea why.  EVDEV_CUSE_DEFAULT_DEVNAME as "input/event1" somehow seems suspect (0-4 already exist?).  I got here because this:

```    539 static void
    540 evdev_enqueue(struct evdev_dev *ed, char *buf, size_t len)
-   541 {
|   542     struct evclient *ec;
|   543
|   544     assert(ed != NULL && buf != NULL);
|   545     assert(len > 0 && len % EVMSG_SZ == 0 && len <= EVBUF_SZ);
|   546
|   547     printf("enqueuing evdev msg_outer\n");
|   548     /* Insert the event message and wakeup all clients. */
|   549     EVDEV_LOCK(ed);
-   550     LIST_FOREACH(ec, &ed->clients, next) {
2   551         EVCLIENT_LOCK(ec);
2   552         evclient_enqueue(ec, buf, len);
2   553         printf("enqueuing evdev msg\n");
2   554         pthread_cond_signal(&ec->cv);
2   555         EVCLIENT_UNLOCK(ec);
2   556     }
|   557     EVDEV_UNLOCK(ed);
|   558
|   559     cuse_poll_wakeup();
|   560 }
```

is showing the "outer" message but not the bare "enqueuing evdev msg\n" in the inside of the loop, leading me to believe no clients are getting attached.  Please help.

**Comment by djvs** [comment]
Also modified the "cuse_methods" provided as evdev_cuse_methods to try to get some insight, with some more log messages.  None of them are called, including evdev_cuse_open, which I guess also may (also?) try to attach an evdev client when it is called.  Code:

```    653 static int
    654 evdev_cuse_open(struct cuse_dev *cdev, int fflags)
-   655 {
|   656     struct evdev_dev *ed = cuse_dev_get_priv0(cdev);
|   657     struct evclient *ec;
|   658     struct hid_interface *hi = ed->cb->get_hid_interface(ed->priv);
|   659
|   660     PRINTE(1, "evdev_cuse_open: cdev(%p) fflags(%#x)\n", cdev,
|   661         (unsigned) fflags);
|   662
|   663     printf("evdev_cuse_open here");
|   664
|   665     /* Initialize a new evdev client. */
|   666     if ((ec = calloc(1, sizeof(struct evclient))) == NULL)
|   667         return (CUSE_ERR_NO_MEMORY);
|   668     ec->evdev = ed;
|   669     if (pthread_mutex_init(&ec->mtx, NULL))
|   670         return (CUSE_ERR_NO_MEMORY);
|   671     if (pthread_cond_init(&ec->cv, NULL))
|   672         return (CUSE_ERR_NO_MEMORY);
|   673     ec->head = ec->tail = ec->buf;
|   674
|   675     cuse_dev_set_per_file_handle(cdev, ec);
|   676
|   677     EVDEV_LOCK(ed);
|   678     ed->refcnt++;
|   679     ec->ndx = ed->totalcnt++;
|   680     ec->enabled = 1;
|   681     LIST_INSERT_HEAD(&ed->clients, ec, next);
|   682     EVDEV_UNLOCK(ed);
|   683
|   684     return (CUSE_ERR_NONE);
|   685 }
```
